Unable to create Fedora releases due to compilation error

Fedora COPR builds changed the warnings they have on or otherwise changed the gcc version and ponyc no longer compiles. We need to fix this to get COPR builds working again.

See: https://copr-be.cloud.fedoraproject.org/results/ponylang/ponylang/fedora-29-x86_64/01042125-ponyc/build.log.

What we need to do:

- [ ] investigate gcc version and flag used
- [ ] setup/modify existing ci to match (not sure which is most applicable)
- [ ] fix the issue

the 2nd 2 steps should be part of the same PR.

I have this error on `gcc (Debian 8.3.0-6) 8.3.0` i686:

```
src/libponyc/pass/refer.c: In function ‘generate_multi_dot_name’:
src/libponyc/pass/refer.c:154:3: error: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
   strncpy(buf + offset, name, slen);
   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/libponyc/pass/refer.c:153:17: note: length computed here
   size_t slen = strlen(name);
                 ^~~~~~~~~~~~
src/libponyc/pass/refer.c:165:5: error: ‘strncpy’ output truncated before terminating nul copying as many bytes from a string as its length [-Werror=stringop-truncation]
     strncpy(buf + offset, name, slen);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/libponyc/pass/refer.c:164:12: note: length computed here
     slen = strlen(name);
            ^~~~~~~~~~~~
cc1: all warnings being treated as errors
make: *** [Makefile:925: build/release/obj-native/libponyc/pass/refer.o] Error 1
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@ All notable changes to the Pony compiler and standard library will be documented
 
 ### Fixed
 
+- Building ponyc with GCC 8+ ([PR #3345](https://github.com/ponylang/ponyc/pull/3345))
 
 ### Added
 

--- a/src/libponyc/pass/refer.c
+++ b/src/libponyc/pass/refer.c
@@ -151,7 +151,7 @@ static const char* generate_multi_dot_name(ast_t* ast, ast_t** def_found) {
   size_t offset = 0;
   const char* name = ast_name(temp_ast);
   size_t slen = strlen(name);
-  strncpy(buf + offset, name, slen);
+  memcpy(buf + offset, name, slen);
   offset += slen;
   temp_ast = ast_parent(temp_ast);
 
@@ -162,7 +162,7 @@ static const char* generate_multi_dot_name(ast_t* ast, ast_t** def_found) {
     temp_ast = ast_sibling(temp_ast);
     name = ast_name(temp_ast);
     slen = strlen(name);
-    strncpy(buf + offset, name, slen);
+    memcpy(buf + offset, name, slen);
     offset += slen;
     temp_ast = ast_parent(temp_ast);
   }
@@ -347,7 +347,7 @@ static const char* suggest_alt_name(ast_t* ast, const char* name)
     // Try with a leading underscore
     char* buf = (char*)ponyint_pool_alloc_size(name_len + 2);
     buf[0] = '_';
-    strncpy(buf + 1, name, name_len + 1);
+    memcpy(buf + 1, name, name_len + 1);
     const char* try_name = stringtab_consume(buf, name_len + 2);
 
     if(ast_get(ast, try_name, NULL) != NULL)

--- a/src/libponyc/pass/verify.c
+++ b/src/libponyc/pass/verify.c
@@ -193,7 +193,7 @@ static const char* get_multi_ref_name(ast_t* ast)
   size_t offset = 0;
   const char* name = ast_name(temp_ast);
   size_t slen = strlen(name);
-  strncpy(buf + offset, name, slen);
+  memcpy(buf + offset, name, slen);
   offset += slen;
   temp_ast = ast_parent(temp_ast);
 
@@ -204,7 +204,7 @@ static const char* get_multi_ref_name(ast_t* ast)
     temp_ast = ast_sibling(temp_ast);
     name = ast_name(temp_ast);
     slen = strlen(name);
-    strncpy(buf + offset, name, slen);
+    memcpy(buf + offset, name, slen);
     offset += slen;
     temp_ast = ast_parent(temp_ast);
   }

--- a/src/libponyc/pkg/package.c
+++ b/src/libponyc/pkg/package.c
@@ -820,7 +820,7 @@ bool handle_path_list(const char* paths, path_fn f, pass_opt_t* opt)
     } else {
       char path[FILENAME_MAX];
 
-      strncpy(path, paths, len);
+      memcpy(path, paths, len);
       path[len] = '\0';
       ok = f(path, opt) && ok;
     }

